echopype/clean/api.py

```diff
@@ -2,74 +2,380 @@
+    func: str, default `nanmean`
+        Pooling function used in the pooled Sv aggregation.
+    depth_bin : str, default `20`m
+        Pooling radius along ``depth`` in meters.
+    num_side_pings : int, default `50`
+        Number of side pings to look at for the pooling.
+    exclude_above : Union[int, float], default `250`m
+        Exclude all depth above (closer to the surface than) this value.
+    transient_noise_threshold : Union[int, float], default `10.0`dB
+        Transient noise threshold value (dB) for the pooling comparison.
+
+    Returns
+    -------
+    xr.Dataset
+        Xarray boolean array transient noise mask.
+
+    References
+    ----------
+    This function's implementation is based on the following text reference:
+
+        Ryan et al. (2015) Reducing bias due to noise and attenuation in
+        open-ocean echo integration data, ICES Journal of Marine Science, 72: 2482–2493.
+
+    Additionally, this code was derived from echopy's numpy single-channel implementation of
+    transient noise masking and translated into xarray code:
+    https://github.com/open-ocean-sounding/echopy/blob/master/echopy/processing/mask_transient.py # noqa
+    """
+    if "depth" not in ds_Sv.data_vars:
+        raise ValueError(
+            "Masking attenuated signal requires depth data variable in `ds_Sv`. "
+            "Consider adding depth with `ds_Sv = ep.consolidate.add_depth(ds_Sv)`."
+        )
+
+    # Copy `ds_Sv`
+    ds_Sv = ds_Sv.copy()
+
+    # Check for appropriate function passed in
+    if func != "nanmean" and func != "nanmedian":
+        raise ValueError("Input `func` is `nanmode`. `func` must be `nanmean` or `nanmedian`.")
+
+    # Warn when `func=nanmedian` since the sorting overhead makes it incredibly slow compared to
+    # `nanmean`.
+    logger.warning(
+        "Consider using `func=nanmean`. `func=nanmedian` is an incredibly slow operation due to "
+        "the overhead sorting."
+    )
+
+    # Setup depth bin
+    ds_Sv, depth_bin = _setup_and_validate(ds_Sv, "depth", depth_bin)
+
+    # Compute pooled Sv
+    pooled_Sv = calc_transient_noise_pooled_Sv(
+        ds_Sv, func, depth_bin, num_side_pings, exclude_above
+    )
+
+    # Compute transient noise mask
+    transient_noise_mask = ds_Sv["Sv"] - pooled_Sv > transient_noise_threshold
+
+    return transient_noise_mask
+
+
+def mask_impulse_noise(
+    ds_Sv: xr.Dataset,
+    depth_bin: str = "5m",
+    num_side_pings: int = 1,
+    impulse_noise_threshold: Union[int, float] = 10.0,
+) -> xr.DataArray:
+    """
+    Locate and create a mask for impulse noise using a ping-wise two-sided comparison.
+
+    Parameters
+    ----------
+    ds_Sv : xarray.Dataset
+        Calibrated Sv data with depth data variable.
+    depth_bin : str, default `5`m
+        Donwsampling bin size along ``depth`` in meters.
+    num_side_pings : int, default `1`
+        Number of side pings to look at for the two-side comparison.
```

Comment:
the paper recommended setting the default to 2 -- check their Table 2. Maybe we use 2?

Reply:
Ah yeah I see that n=2 note below their table 2. I'll use that insetad.